### DISCUSSION TITLE
Mapped fields in useFieldArray lose value on append, remove, or submit

### DISCUSSION THREAD

**Discussion by hellocng** [react-hook-form/react-hook-form#4007] [Q&A] [answered]
This is probably not the best way to use RHF or field arrays so if anyone has any suggestion on a different way to implement, I'm open to ideas.

I have a list of items, and they are buttons to add themselves to an order. This is like a point-of-sales application so this is the most efficient way, instead of filling out a form. Each of the item may have some options, and these options are different between the items. Ultimately, I want to store the option ids inside an object.

When I append, the form resets and all values are lost. Sometimes a render loop happens with this error:
`
index.js:1 Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.`

With just one item, the form works as expected. Sometimes, I can somehow append a few items, change their values and submit. Once I submit, render loop happens again but I can see that the data went through fine.

[CSB](https://codesandbox.io/s/eloquent-chaplygin-vo3se?file=/src/form.js)

**Comment by hellocng** [comment]
So I narrowed down the issue to even time the form is appended or submitted, the fields are taking values from defaultValue again. My hackery approach is to just set state and use it for defaultValue:

`const [optionValues, setOptionValues] = useState({});`

`<FormControlLabel
                    label={option.name}
                    control={
                      <Controller
                        name={`orderItems[${index}].optionId.${option.id}`}
                        control={control}
                        defaultValue={
                          optionValues[
                            `orderItems[${index}].optionId.${option.id}`
                          ]
                        }
                        render={({ onChange, value, ref }) => (
                          <Checkbox
                            onChange={(e) => {
                              onChange(e.target.checked);
                              setOptionValues({
                                ...optionValues,
                                [`orderItems[${index}].optionId.${option.id}`]: e
                                  .target.checked
                              });
                            }}
                            checked={value}
                            inputRef={ref}
                            color="primary"
                          />
                        )}
                      />
                    }
                  />`

[CSB](https://codesandbox.io/s/eloquent-chaplygin-vo3se?file=/src/form.js)

Any better ideas?

**Reply by bluebill1049** [reply]
why do you need this line?  (probably the root cause)

`const [optionValues, setOptionValues] = useState({});`

**Reply by hellocng** [reply]
Actually, that is the solution.

Before setting the value in state, every append will reset the form, which means all the checkboxes go back to defaultValues.

So now, I set the value as the defaultValue so after every append, I will get the value back when the form resets. I’m guessing the form resets because map is creating new array and new fields. Not sure if we can get the new fields to get values from the old field with the same name.

As for the render loop in the title, the solution was adding ‘shouldUnregister =false’.

**Reply by bluebill1049** [reply]
> every append will reset the form? 

If you have supplied defaultValue they shouldn't: https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv

**Reply by hellocng** [reply]
This is how the form was [previously](https://codesandbox.io/s/stoic-cohen-mqh6i?file=/src/form.js). The defaultValue was `false`.

In the original [CSB](https://codesandbox.io/s/eloquent-chaplygin-vo3se?file=/src/form.js), the new issue is the warning of not having a defaultValue set, even though I do. But, at least it is working.

**Comment by hellocng** [comment]
Update to solution:

After more testing, I found that if I remove an item from the field array, the subsequent items will need new values/defaultValues as the index of the items are stored as strings in the keys of the object in state. To fix that, I loop through all of the keys to subtract 1 from all keys with indexes greater than the item removed. So now, I think my implementation is fully working but again, the main issue is that if the fields were generated from map, append/remove/submit cause the fields to lose their values because the whole field array becomes a new field array.

[Broken implementation CSB](https://codesandbox.io/s/stoic-cohen-mqh6i?file=/src/form.js)

[Working implementation CSB - monkey patching](https://codesandbox.io/s/eloquent-chaplygin-vo3se?file=/src/form.js)

**Comment by hellocng** [comment]
The temporary solution has become impractical to continue using as I have to manually manage the form state of each field I add to the form. Can someone please take a look at this [CSB](https://codesandbox.io/s/stoic-cohen-mqh6i?file=/src/form.js)? Both the CheckBoxes and the TextFields lose their value when I append, remove, or submit.

**Comment by hellocng** [comment, marked as answer]
I have finally figured out the root of the problem. I was nesting component declaration as I tried to create a component inside the form component to put into the TextField input component. Every time something in Orders change, React creates a new function and therefore new component, unmounting and remounting a new html element. For whoever may have this same problem in the future, here is the final [CSB](https://codesandbox.io/s/busy-williams-ohkq1?file=/src/form.js).

**Reply by bluebill1049** [reply]
good find, if you working closely with useFieldArray I think it's worth checking out V7. https://twitter.com/HookForm/status/1357882714439839751
